File: ADT/trie/208.py
class Trie:
    class TrieNode:
        def __init__(self):
            self.children = [None] * 26
            self.isEndOfWord = False

    def __init__(self):
        self.root = self.TrieNode()
        

    def insert(self, word: str) -> None:
        crawler = self.root
        for ch in word:
            index = ord(ch) - ord('a')
            if crawler.children[index] is None:
                crawler.children[index] = self.TrieNode()
            crawler = crawler.children[index]
        crawler.isEndOfWord = True
        

    def search(self, word: str) -> bool:
        crawler = self.root
        for ch in word:
            index = ord(ch) - ord('a')
            if crawler.children[index] is None:
                return False
            crawler = crawler.children[index]
        return crawler is not None and crawler.isEndOfWord
        

    def startsWith(self, prefix: str) -> bool:
        crawler = self.root

        for ch in prefix:
            index = ord(ch) - ord('a')
            if crawler.children[index] is None:
                return False
            crawler = crawler.children[index]

        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: ADT/trie/208.js
class TrieNode {
    constructor() {
        this.children = new Array(26).fill(null);
        this.isEndOfWord = false;
    }
}

class Trie {

    constructor() {
        this.root = new TrieNode();
    }

    insert(word) {
        let crawler = this.root;

        for (let ch of word) {
            let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
            if (crawler.children[index] === null) {
                crawler.children[index] = new TrieNode();
            }
            crawler = crawler.children[index];
        }

        crawler.isEndOfWord = true;
    }

    search(word) {
        let crawler = this.root;

        for (let ch of word) {
            let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
            if (crawler.children[index] === null) {
                return false;
            }
            crawler = crawler.children[index];
        }

        return crawler !== null && crawler.isEndOfWord;
    }

    startsWith(prefix) {
        let crawler = this.root;

        for (let ch of prefix) {
            let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
            if (crawler.children[index] === null) {
                return false;
            }
            crawler = crawler.children[index];
        }

        return true;
    }
}
